text/0018-bee-bundle/0018-bee-bundle.md

```diff
@@ -0,0 +1,281 @@
+
+For these reasons, transactions have to be processed as a whole, in groups called bundles. A bundle is an atomic operation in the sense that either all or none of its transactions are accepted by the network. Even single transactions are propagated through the network within a bundle making it the only confirmable communication unit of the IOTA protocol.
+
+This RFC proposes ways to create and manipulate a bundle and describe the associated algorithms.
+
+Useful links:
+- [What is a bundle?
+](https://docs.iota.org/docs/getting-started/0.1/introduction/what-is-a-bundle)
+- [Bundles and transactions
+](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+
+# Motivation
+
+<!-- TODO -->
+
+# Detailed design
+
+<!-- TODO -->
+
+## Bundle and BundleBuilder
+
+Transactions are final and bundles, essentially being arrays of transactions, are also final. Once a bundle is created and validated, it shouldn't be tempered. For this reason we have a `Bundle` type and a `BundleBuilder` type. An instantiated `Bundle` object represents a syntactically and semantically valid IOTA bundle and a `BundleBuilder` is the only gateway to a `Bundle` object.
+
+### Bundle
+
+As bundles are final, they shouldn't be modifiable outside of the scope of the bundle module.
+
+A bundle can simply be represented as an array of transactions:
+
+```rust
+struct Bundle {
+    transactions: Vec<Transaction>
+}
+```
+
+And its implementation should only allow to retrieve transactions:
+
+```rust
+impl Bundle {    
+    pub fn transactions(&self) -> &Vec<Transaction> {
+        &self.transactions
+    }
+}
+```
+
+### BundleBuilder
+
+The `BundleBuilder` offers a simple and convenient way to build bundles:
+
+```rust
+pub struct BundleBuilder {
+    transactions: Vec<TransactionBuilder>
+}
+```
+
+```rust
+impl BundleBuilder {    
+    pub fn hash(&self) {
+      unimplemented!()
+    }
+
+    pub fn finalise(&self) {
+      unimplemented!()
+    }
+
+    pub fn sign(&self) {
+      unimplemented!()
+    }
+
+    pub fn pow(&self) {
+      unimplemented!()
+    }
+
+    pub fn validate(&self) {
+      unimplemented!()
+    }
+
+    pub fn build(&self) {
+      unimplemented!()
+    }
+}
+```
+
+*We do not list parameters and/or return values as they are implementation details.*
+
+## Algorithms
+
+In this section, we describe the algorithms needed to build a `Bundle`. The lifecycle of a `BundleBuilder` depends on if it's being used in client side or server side:
+- client side: `finalise` -> [`sign` ->] -> [`pow` ->] -> `validate` -> `build`
+- server side: `addTransaction`/`addTransactionBuilder` -> `validate` -> `build`
+
+*`sign` is optional because data transactions don't have to be signed. `pow` is optional because one can use remote pow instead.*
```

Comment:
As we introduce intermediate type, I think pow validation will be mandatory before built? Besides, even it's remote pow, client still needs to wait for returned nonce no?